Add property set/get hooks.

wt opened the pull request:
This change adds a number of property modifiers that add the ability to
specify a hook functions to run before and/or after setting and/or
getting the values of properties.

wt commented:
@toasteater I think this addresses the comments you made before.

toasteater commented on gdnative-derive/src/native_script.rs:
```diff
@@ -36,12 +36,54 @@ pub(crate) fn derive_native_class(input: TokenStream) -> TokenStream {
                 None
             };
 
+            let get_prehook = match config.get_prehook {
+                Some(x) => {
+                    let expr_str = format!("{}(this, _owner)", x);
+                    let expr = syn::parse_str::<Expr>(expr_str.as_str());
+                    expr.ok()
+                }
+                None => None,
+            };
```
Try using `Option::map` and `syn::parse_quote!`?

```rust
let get_prehook = config.get_prehook.map(|path| parse_quote!(#path(this, _owner)));
```

toasteater commented on gdnative-derive/src/native_script/property_args.rs:
```diff
@@ -1,12 +1,20 @@
 pub struct PropertyAttrArgs {
     pub path: Option<String>,
     pub default: Option<syn::Lit>,
+    pub get_prehook: Option<String>,
+    pub get_posthook: Option<String>,
+    pub set_prehook: Option<String>,
+    pub set_posthook: Option<String>,
```
Perhaps `Option<syn::Path>` here because it's better to verify that these are actually paths that can be called? I'm also not very sure about the naming. I'd prefer `before_*` and `after_*`, since they're easier to understand than `_prehook` and `_posthook`.

wt replied:
Verifying that's it's a path seems fine. How would you feel about taking it one step further and generating and Option<ExprCall> for each one that is the actual function call?

toasteater replied:
I'd prefer separating option parsing and code generation, so it's cleaner to change the implementation later if necessary.

wt replied:
I am just backing off on that. The path idea seems pretty good. I'll see what I can do with it.

wt commented:
In the newest version, I have removed the get hook since it's can't mutate the object. I also renamed all the args to be more in line with your suggestion. However, because there is no after get hook, I made that one `get_hook`. For symmetry, I called the before and after for the setter `set_before_hook` and `set_after_hook`. I'm open to better names if you don't like them.

I think I have addressed your comments. PTAL, and see what you think.

toasteater commented:
`set_before_hook` and `set_after_hook` are ambiguous since they can be interpreted as the actions of "setting the before/after hook". It is possible for a get hook to mutate the script struct using interior mutability, and we can also generate `&mut` getters instead if get hooks are present, so I'd say that an `after_get` hook might still be useful.

toasteater commented on gdnative-derive/src/native_script.rs:
```diff
@@ -36,12 +36,40 @@ pub(crate) fn derive_native_class(input: TokenStream) -> TokenStream {
                 None
             };
 
+            let get_hook = if let Some(get_hook_expr) = config.get_hook {
+                let get_hook: ExprCall = parse_quote!(#get_hook_expr(this, _owner));
+                Some(get_hook)
+            } else {
+                None
+            };
```
What's stopping you from using `Option::map`?

wt replied:
> `set_before_hook` and `set_after_hook` are ambiguous since they can be interpreted as the actions of "setting the (unspecified) before/after hook". It is possible for a get hook to mutate the script struct using interior mutability, and we can also generate `&mut` getters instead if get hooks are present, so I'd say that an `after_get` hook might still be useful.

I will add back the before/after get hooks and rename everything appropriately. I am not going to be working on getting doing the mut getters. That seems like a separate change. Unfortunately, I don't have time to grapple with that one for now.

wt replied:
This part is done. Working on other comments.

wt replied:
This commend landed in a weird place. Let me fix this.

wt replied:
My inexperience with rust. I'm working on this right now.

wt replied:
Now fixed.

toasteater commented on gdnative-derive/src/native_script/property_args.rs:
```diff
@@ -37,6 +43,57 @@ impl<'a> Extend<&'a syn::MetaNameValue> for PropertyAttrArgsBuilder {
+                        panic!("get_hook value is not a string literal");
+                    };
+
+                    if let syn::Expr::Path(path_expr) =
+                        syn::parse_str::<syn::Expr>(string.as_str()).expect("Invalid syntax.")
+                    {
+                        if let Some(old) = self.get_hook.replace(path_expr.path) {
+                            panic!("there is already a get_hook value set: {:?}", old);
+                        }
+                    } else {
+                        panic!("Not a path expression.")
```
Why parse as a path expression? `syn::Path` itself also implements `Parse`. You can just parse it directly.

wt replied:
I am working on this now. My theory was that having the expression might allow better error messages. However, I can't really return errors from the extend method, so I guess that doesn't matter.

FWIW, this is probably something that should be fixed in the future. Implementing the Extend trait doesn't make a ton of sense to me since it doesn't seem to give us anything here. And it would be super useful to be able to return errors from that function so that we could have better compiler error messages. For example, when I was getting the ExprPath instead, it has all the spans that I could use to generate a good syn:Error so that the compiler could show the bad code instead of only the macro invocation, which is harder than it needs to be to debug.

wt replied:
Okay. I have addressed this.

wt commented:
> `set_before_hook` and `set_after_hook` are ambiguous since they can be interpreted as the actions of "setting the (unspecified) before/after hook". It is possible for a get hook to mutate the script struct using interior mutability, and we can also generate `&mut` getters instead if get hooks are present, so I'd say that an `after_get` hook might still be useful.

I will add back the before/after get hooks and rename everything appropriately. I am not going to be working on getting doing the mut getters. That seems like a separate change. Unfortunately, I don't have time to grapple with that one for now.

This now done.

toasteater commented:
Ok, I think this implementation looks fine for now. Could you add some tests for it?

wt commented:
I know this is gonna be a dumb questions. But where are the tests. I don't see any tests in the derive crate.

karroffel commented:
It doesn't have to be a test in the derive crate, it should be fine to add a test some where in the [`test` directory](https://github.com/godot-rust/godot-rust/tree/master/test/src). For example [this file](https://github.com/godot-rust/godot-rust/blob/master/test/src/test_derive.rs) has some tests regarding derive macros.

Nice PR btw! :sparkles:

wt commented:
I need some help with the testing. I have got things mostly wired up. I know the change it working from using it, but I am not sure how to setup the test rig properly as I essentially need to wire up an object as it would be used in godot and this get the properly. I have created the script (a struct that derives from NativeClass) and the owner object (the Node in this case). However, I am not sure how to wire them together to allow a call to `owner.get("fake_property")` to by interacting with the nativeclass object. Can anyone help? I have posted the new code. please checkout the test in my changes.

toasteater commented:
Actually the code posted above might not work due to borrowing rules. This should work when the script is registered:

```rust
        use gdnative::nativescript::user_data::Map;

        let thing = Instance::<FakeNativeClass, _>::new();
        let (owner, script) = thing.decouple();
        owner.set("simple_bool", true);
        script.map(|script| {
            assert_eq!(script.callback_calls, 1);
        }).unwrap();
        owner.free();
```

wt commented:
All of those are failing with this message:
```
 -- test_derive_nativeclass_with_property_before_set
thread '<unnamed>' panicked at 'assertion failed: `(left != right)`
  left: `0x0`,
 right: `0x0`: script instance should not be null', /home/wt/projects/godot/godot-rust/gdnative-core/src/nativescript/class.rs:234:13
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

Why would the script be null?

wt commented:
Oh, duh, you said I had to register the script. How do I do that? I'm assuming this is like the init function that adds classes. Is there a way to isolated the init calls?

karroffel commented:
You would have to put your type in the "file" scope, and then you can use the [`register` function in the file](https://github.com/godot-rust/godot-rust/blob/428a47a5fdfc346d3a34ee40aa23b3584812e253/test/src/test_derive.rs#L13) to register your type.

Here is how another test does it:
https://github.com/godot-rust/godot-rust/blob/428a47a5fdfc346d3a34ee40aa23b3584812e253/test/src/test_free_ub.rs#L13-L15

wt commented:
Thanks to all your help. I got it. I am finishing it up right now.

wt commented:
I just pushed a test to make sure that all the different hook types are working. I think it's probably ready to land unless we find something else.

toasteater commented:
> I wonder if we could move some of this to unit test that can be run with cargo test

No, because all the tests here require a live instance of Godot.

toasteater commented on test/src/test_derive.rs:
```diff
@@ -160,3 +167,181 @@ fn test_derive_owned_to_variant() -> bool {
+
+#[derive(gdnative::NativeClass)]
+#[inherit(Node)]
+struct FakeNativeClassWithPropertyHooks {
+    #[property(before_set = "Self::before_get_fake_callback")]
+    before_get_bool: bool,
+    #[property(after_set = "Self::after_get_fake_callback")]
+    after_get_bool: bool,
+    #[property(before_set = "Self::before_set_fake_callback")]
+    before_set_bool: bool,
+    #[property(after_set = "Self::after_set_fake_callback")]
+    after_set_bool: bool,
+
+    pub before_get_callback_calls: u32,
+    pub after_get_callback_calls: u32,
+    pub before_set_callback_calls: u32,
+    pub after_set_callback_calls: u32,
+}
```
I think these names are a bit too verbose. Maybe instead of `FakeNativeClassWithPropertyHooks`, just make it `PropertyHooks`? Same with `Self::before_get_fake_callback` and `before_get_callback_calls` -- just make that `Self::before_get` and `before_get_calls`.

toasteater commented on test/src/test_derive.rs:
```diff
@@ -160,3 +167,181 @@ fn test_derive_owned_to_variant() -> bool {
+        }
+    }
+
+    fn before_get_fake_callback(_script: &mut Self, _owner: TRef<Node, Shared>) {
+        _script.before_get_callback_calls += 1;
+    }
+
+    fn after_get_fake_callback(_script: &mut Self, _owner: TRef<Node, Shared>) {
+        _script.after_get_callback_calls += 1;
+    }
+
+    fn before_set_fake_callback(_script: &mut Self, _owner: TRef<Node, Shared>) {
+        _script.before_set_callback_calls += 1;
+    }
+
+    fn after_set_fake_callback(_script: &mut Self, _owner: TRef<Node, Shared>) {
+        _script.after_set_callback_calls += 1;
+    }
```
Underscores are put in front of identifiers to silence unused variable warnings. They are not mandatory. Also since this is in an impl block, you can use `&mut self` instead of `_script: &mut Self`.

toasteater commented on test/src/test_derive.rs:
```diff
@@ -160,3 +167,181 @@ fn test_derive_owned_to_variant() -> bool {
+    }
+}
+
+fn test_derive_nativeclass_with_property() -> bool {
+    println!(" -- test_derive_nativeclass_with_property");
+
+    #[derive(gdnative::NativeClass)]
+    #[inherit(Node)]
+    struct FakeNativeClass {
+        #[property]
+        pub simple_bool: bool,
+    }
+    impl FakeNativeClass {
+        fn new(_owner: &Node) -> Self {
+            Self { simple_bool: false }
+        }
+    }
+
+    let ok = std::panic::catch_unwind(|| {
+        let owner = Node::new();
+        let class = FakeNativeClass::new(&owner);
+        assert_eq!(class.simple_bool, false);
+    })
+    .is_ok();
+
+    if !ok {
+        gdnative::godot_error!("   !! Test test_derive_nativeclass_with_property failed");
+    }
+
+    ok
+}
```
This test is useless because it's simply accessing the Rust field, not through GDNative. `#[property]` without any hooks is a pretty trivial case and I don't think it really needs to be tested, so you may also remove it outright.

wt replied:
My theory here was testing this built a foundation that allowed me to trust the rest of the testing. What's the harm in keeping it? In fact, during my dev, a similar manually executed test caught errors I had a made a number of times. I would prefer to keep it for the coverage. Having said that, if you are strongly opposed, I can remove it.

toasteater replied:
Did you read the comment? I said that it doesn't test what you think it's testing. What is tested here by this code is just rustc's ability to calculate field offsets, and not the implementation of `#[property]` **at all**. If you want to keep it, then at least use `Instance` and `Object::set`.

wt replied:
I did read your comment, and I did understand it. What I was testing was that the property didn't disappear because of the #[property]. I agree that using the something like a set and a get would be better, so I will add that.

FWIW, your response reads like you're annoyed. I am sorry if I annoyed you. I just want a good quality change. And I appreciate your help.

toasteater commented on test/src/test_derive.rs:
```diff
@@ -160,3 +167,181 @@ fn test_derive_owned_to_variant() -> bool {
+    ok
+}
+
+fn test_derive_nativeclass_with_property_hooks(
+    set_function: fn(Ref<Object, Unique>) -> Ref<Object, Unique>,
+    expected_counts: &[u32; 4],
+) -> bool {
+    std::panic::catch_unwind(|| {
+        use gdnative::nativescript::user_data::Map;
+
+        let thing = Instance::<FakeNativeClassWithPropertyHooks, _>::new();
+        let (owner, script) = thing.decouple();
+        let owner = set_function(owner.upcast::<Object>());
+        script
+            .map(|script| {
+                assert_eq!(script.before_get_callback_calls, expected_counts[0]);
+                assert_eq!(script.after_get_callback_calls, expected_counts[1]);
+                assert_eq!(script.before_set_callback_calls, expected_counts[2]);
+                assert_eq!(script.after_set_callback_calls, expected_counts[3]);
+            })
+            .unwrap();
+        owner.free();
+    })
+    .is_ok()
+}
```
Although this does verify that the hooks are called at all, I don't know if this is the best way to test them because the order of operation isn't tested. Maybe make the hooks increment the values depending on the property value? So you may verify whether the hook is actually called before/after the property is set by setting it to a different value.

wt replied:
I thought about doing something like 1, 2, 4, and 8 calls to each to verify it all in one test. However, I wanted this as more of a unit test of each. I can add another test to do that if you want, but i really think these tests should stay so that breaking any one of the callback paths it detected by an independent test. Having narrow tests is something I believe in so that when someone breaks them (which will inevitably happen), it's easy to find what was broken. Especially with this test harness, I don't want to make it any harder.

toasteater replied:
Did you read the comment? I believe that the order of operation is part of the unit here, but unit test or not, it isn't tested! How would this code expose the problem if someone reversed, for example, `#before_set` and `#after_set` in the proc-macro code? Of course it won't.

wt replied:
Yeah. I did read your comment, and I think I've been trying to implement your ideas, and I probably misinterpreted it, and I do appreciate your feedback. However, you really being quite harsh at this point. I am not going to be getting back to this this weekend. I'll see what i can do next weekend when I have more time.

toasteater replied:
I see. Would you mind if I made a PR based on this, then?

wt replied:
If you want to take it to completion, that's fine with me. FWIW, I'm not saying I won't do it. I just won't have time this weekend (responsibilities at home), or next week while I am at work. Alternatively, I would also support just landing this and having another change that refines the testing, whichever is easier for you and the project.

Having said all this, I am going to provide one piece of unsolicited feedback. I worked pretty hard on this change to make sure it was acceptable to this project. Your communication was pretty harsh at certain points. FWIW, it made me feel like my work wasn't that interesting to the project. Please don't assume that I was maliciously not reading your comments, I was trying to act on them in good faith. If anything, I am just not as experienced with rust as you, and I just didn't fully grok all of them at first. I could imagine folks that might be chased away by such harsh reactions.

Having said all that, thank you again for you help. I'm certain this change would not be as good as it is without your input.

toasteater replied:
That's great to hear, and thanks for your *unsolicited feedback*. *FWIW*, please allow me to give you a piece of that too:

Provided that you claim to *know* that you're not very experienced with Rust, and since you're new to the project, have it ever occurred to you that maybe, perhaps, with any slim possibility, that your ideas might not be as informed as you think, and you shouldn't stick too stubbornly to them? I'm not usually this harsh to people, and I wasn't this harsh to you in the beginning. You're correct to observe that I'm annoyed, but have you wondered *why*? *FWIW*, this piece of *unsolicited feedback* certainly didn't help.

Thanks for your efforts regardless. I wish you a good day.

wt replied:
You know. I just read through the whole conversation through both PRs. I can definitely understand how I might have seemed that I was more set into some of the ideas than I was. I am sorry for that. However, my intention was literally to ask more questions and get the information I needed to make this change acceptable. I even dumped my entire first PR after asking some questions to get a better understanding of what you wanted. In fact, I am pretty sure I ended up changing everything you asked for. From my perspective, I wasn't rejecting your feedback. I was trying to get more information and incorporate it. Can you please help me understand how I could have better communicated that better?

toasteater replied:
Ok. I'll bite and assume you're being sincere, but maybe just consider what your actions will look like from the perspective of a maintainer before you do them? I believe it's mostly common sense but:

If you didn't understand a comment, just ask kindly and don't go to extra lengths to dismiss the things that you *didn't even understand*. There is a [development Discord](https://discord.gg/FNudpBD) in `CONTRIBUTING.md`, with me and many other contributors. We'd all be glad to help if you just asked kindly.

Discuss changes if you don't like what's proposed. And yes, if me (or someone else) post mock code for a "provisional API", *that counts as a proposal for what the API will look like*, e.g. identifiers and things. Don't just dismiss them without carrying out the *entirety* of the proposed changes. Again, if you can't do it, ask kindly for help *before you dismiss the conversation*.

I'm just a volunteer working on this project in my free time. I'm not paid to serve as your subordinate. Sadly, there's a limit to how kind I can be towards someone who comes off as constantly trying to sound like a superior. If you would like to ever contribute again in the future, please try to treat people kindly, and you should get the same in return.

toasteater commented:
Closing in favor of #585.

karroffel commented:
I think that @wt has been treated unfairly harsh here. I share wt's impression that "I could imagine folks that might be chased away by such harsh reactions."

Meaning it with the very best intentions, I personally had a feeling that wt isn't that experienced with Rust yet. The code that I saw that from was still pretty clear, but not *idiomatic*. And that's totally cool. Making a PR on a project where you are not familiar with the internals yet, written in a language you don't have the most experience in yet is already very commendable and very welcome. Even if a PR wouldn't end up merged, some other person might still find an open (or closed) PR useful. I have used patches from closed PRs before because they did solve my issue.

When somebody less experienced makes a PR that can mean that a bit more guidance is needed, but that's to be expected, not everybody has the same expertise. This is all open source work, nobody here has any obligation to do anything, be it making a PR or reviewing a PR. In my view, wt was not *demanding* supervision **or** to have changes accepted as they were. From what I see, most feedback has either been addressed or is still "in discussion".

Everybody here does things in their freetime, if the communication gets in the way, things will just stop eventually. Nobody is getting paid here to endure having a bad time.

In particular to @toasteater, if you are getting annoyed and are about to write a comment like [this](https://github.com/godot-rust/godot-rust/pull/581#discussion_r484023132), I invite you to *not* that send comment and message me instead and ask if I can take over. No need for you to get into a bad mood, and even less need to make another person receive an angry text.
